### PR TITLE
Store build logs in S3

### PR DESCRIPTION
**Pull request by Nemo157**
This is a first step on the path to #787 

The first 4 commits are refactoring, adding tests for the builds list and build logs and splitting them into separate handlers and templates (I don't think there's much benefit to having the list of other builds below the log, and it makes the code simpler to reason about).

The last commit adds support for rendering the build log from either database or S3, and changes the build process to upload to S3.

The next step would be to migrate existing build logs from the database to S3 so we can delete the code checking for both. Then thread the build logs for other targets through and upload and render them all.

We could also consider bumping the max build log size up, which would allow reverting https://github.com/rust-lang/docs.rs/pull/1192 if having verbose logs would be useful.

### CODE REVIEW

File: src/storage/mod.rs
Comment (jyn514) [inline src/storage/mod.rs:276]: Any reason to change this?

File: src/test/fakes.rs
Comment (jyn514) [inline src/test/fakes.rs:332]:
```suggestion
        let default_target = self.default_target.unwrap_or(docsrs_metadata::HOST_TARGET);
```

File: src/web/build_details.rs
Comment (jyn514) [inline src/web/build_details.rs:182]: That seems fine, yeah - the alternative is to check each time we get something from the database whether there's also something in S3, which is expensive and shouldn't be necessary. Maybe we could add a mode that double checks it's deduplicated but that also sounds expensive.

**Review by jyn514 — COMMENTED** [review]
How does this behave for multiple builds? Will it show/store the additional logs?

**Comment by jyn514** [comment]
> The next step would be to migrate existing build logs from the database to S3 so we can delete the code checking for both. Then thread the build logs for other targets through and upload and render them all.

We should decide if we want to support running docs.rs in prod without the database (we shouldn't be special casing build logs when documentation can still be stored in the DB). Since we have min.io for testing locally and all our data in prod is in S3, I guess there's no real reason to keep the DB support around? @pietroalbini what do you think?

File: src/test/fakes.rs
Comment (jyn514) [reply src/test/fakes.rs]: Never mind, you make changes in a later commit that make my suggestion pointless.

**Comment by jyn514** [comment]
>  the test failed: failed to get a database connection
  caused by: timed out waiting for connection: db error: FATAL: sorry, too many clients already

Ugh, I'll review https://github.com/rust-lang/docs.rs/pull/1216.

Do you know why it only failed here? It's been working fine on other PRs ...

**Comment by Nemo157** [comment]
> > The next step would be to migrate existing build logs from the database to S3 so we can delete the code checking for both. Then thread the build logs for other targets through and upload and render them all.
> 
> 
> 
> We should decide if we want to support running docs.rs in prod without the database (we shouldn't be special casing build logs when documentation can still be stored in the DB). Since we have min.io for testing locally and all our data in prod is in S3, I guess there's no real reason to keep the DB support around? @pietroalbini what do you think?

Well, I say S3, but I actually mean "the configured storage", so in tests they're still stored in the DB, just in the files table instead.

**Comment by Nemo157** [comment]
> >  the test failed: failed to get a database connection
> 
>   caused by: timed out waiting for connection: db error: FATAL: sorry, too many clients already
> 
> 
> 
> Ugh, I'll review https://github.com/rust-lang/docs.rs/pull/1216.
> 
> 
> 
> Do you know why it only failed here? It's been working fine on other PRs ...

3 new web tests may have just put it over the limit.

File: src/web/builds.rs
Comment (Nemo157) [reply src/web/builds.rs:224]: They're returned `BY id DESC`, so the most recent build is at the top of the list.

File: src/web/builds.rs
Comment (Nemo157) [reply src/web/builds.rs:257]: Yep.

File: src/web/routes.rs
Comment (Nemo157) [reply src/web/routes.rs:88]: Yes, it's one build per page. It no longer shows the list of other builds and limits below the build log, which means it doesn't need the additional code for loading that data in this handler too. You can just hit back to go back to the list.

It doesn't drastically simplify the code right now, but I think it will help a lot once the `build_details_handler` needs to start querying the list of different targets to show.

File: src/storage/mod.rs
Comment (Nemo157) [reply src/storage/mod.rs:154]: It's not accurate anymore, the storage selection is controlled by a different environment variable and I don't think it's something that really should be documented on this method, it's a fundamental part of how `Storage` works.

File: src/storage/mod.rs
Comment (Nemo157) [reply src/storage/mod.rs:226]: There's no reason to ever take `Iterator` if you can take `IntoIterator` instead (a.k.a. you're not partially consuming the iterator), it's strictly more general and supports everything `Iterator` does. I needed it with a past implementation of `store`, but after changing that around I guess the original signature would work.

File: src/test/fakes.rs
Comment (Nemo157) [reply src/test/fakes.rs]: (personally I dislike directly ever forwarding results in anything other than direct delegation methods; always using `?` is more consistent, highlights where the errors can come from, and simplifies diffs).

File: src/storage/mod.rs
Comment (Nemo157) [reply src/storage/mod.rs:276]: It allows passing the `&String` directly in `store_one` instead of having to manually wrap it in a `Path`.

File: src/docbuilder/rustwide_builder.rs
Comment (Nemo157) [reply src/docbuilder/rustwide_builder.rs:432]: Hmmm, my plan was to delete the column from the database post-migration. Do we expect to ever have this actually fail?

File: src/web/builds.rs
Comment (pietroalbini) [reply src/web/builds.rs:172]: That endpoint is [used by crates.io](https://github.com/rust-lang/crates.io/blob/82cb342d982858de487dba6370a09fdea0ba44c0/app/controllers/crate/version.js#L153).

**Comment by pietroalbini** [comment]
> We should decide if we want to support running docs.rs in prod without the database (we shouldn't be special casing build logs when documentation can still be stored in the DB). Since we have min.io for testing locally and all our data in prod is in S3, I guess there's no real reason to keep the DB support around?

This doesn't use S3 directly, it uses `Storage`, which also has a database backend.

File: src/docbuilder/rustwide_builder.rs
Comment (jyn514) [reply src/docbuilder/rustwide_builder.rs]: Well, I think we should treat it the same as failing to upload documentation. And looking at the code we ... give up on the first failure :facepalm: so it could be inconsistent at any point if the builder loses internet for some reason.

So I think there's room for improvement here, but it's probably a large change and shouldn't block this PR.

File: src/web/routes.rs
Comment (jyn514) [reply src/web/routes.rs:92]: Hmm, my concern was that people would no longer have the limits staring them in the face when they see a failed build, but it looks like the 'build failed' link takes you to /builds directly, so you do see the limit table before seeing the log itself. I'm still slightly worried people won't think of it, but it's not a big deal, I'd be ok either way.

**Review by jyn514 — APPROVED** [review]

File: src/docbuilder/rustwide_builder.rs
Comment (Nemo157) [reply src/docbuilder/rustwide_builder.rs]: I actually ended up changing this in my next steps preparing to support all the logs, I think we can capture almost any error that happens during the build and attempt to shove it into the database.

**Pull request merged**
